### DISCUSSION TITLE
Custom executor service in Caffeine cache (AsyncLoader)

### DISCUSSION THREAD

**Discussion by ankushjindal81** [ben-manes/caffeine#508] [Q&A] [answered]
I am using Caffeine AsyncLoader to pull some data from third party apis, which take some time. The problem I am facing is Caffeine AsyncLoader uses ForkJoin common pool and I want to pass the previous thread's context to the function which is loading the data. For that I need to use custom executor service in caffeine, so that caffeine can use custom executor service for async loading. Is there any way to do it?

**Comment by ben-manes** [comment, marked as answer]
You can set `Caffeine.executor(executor)` in your builder, and it will use that for all calls. Does that help?

**Reply by ankushjindal81** [reply]
Sure Let me check. Thanks a lot for quick reply.

**Reply by ben-manes** [reply]
Did this solve your problem or are you still having issues?

**Reply by ankushjindal81** [reply]
It solved my problem. Thanks :)

**Reply by formanojhr** [reply]
@ben-manes a specific question for you on this topic. If we pass an executor with a bounded queue we do need to specific a rejectionhandler. There is a comment in the executor saying that will lead to non deterministic behavior. Does that mean we cannot bound the worker pool for the executor passed in ?

**Reply by ben-manes** [reply]
The original documentation was pessimistic due to concern of unknown bugs. It stated,
> Beware that configuring a cache with an executor that throws {@link RejectedExecutionException} may experience non-deterministic behavior.

This was clarified in 3.0.5 and currently documented as,
> Beware that configuring a cache with an executor that discards tasks or never runs them may experience non-deterministic behavior.

The cache should handle any exception being thrown, like `AbortPolicy` which throws `RejectedExecutionException`. The exception will either propagate that to the caller (e.g. `cache.get(key)`) or the task will be called directly if an internal dispatch (e.g. `removalListener.onRemoval(k, v, cause)`). If the executor uses `DiscardPolicy` or similarly never runs the task then the cache can only compensate for its own maintenance work (e.g. size eviction). For any other calls like removal notifications, refresh, or async loads then it can't assist in handling those future tasks. For cache loads this would leaves incomplete entries floating in the cache since in-flight futures are not evictable. The documentation is now about warning against misuse that violates the application's own expectations.

In short, it should all work as you would reasonably expect. The library has matured so there are few rough spots and the tone softened to mirror that.

**Reply by formanojhr** [reply]
Thanks for the prompt response @ben-manes
